src/charm.py

```diff
@@ -1,117 +1,278 @@
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_available, self._on_certificate_available
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_expiring, self._on_certificate_expiring
+        )
+        self.framework.observe(
+            self.certificates.on.certificate_invalidated, self._on_certificate_invalidated
+        )
+        self.framework.observe(self.on.get_certificate_action, self._on_get_certificate_action)
+        self.framework.observe(
+            self.certificates.on.all_certificates_invalidated,
+            self._on_all_certificates_invalidated,
+        )
+
+    def get_hostname(self) -> str:
+        """Get a list containing all ingress hostnames.
+
+        Returns:
+            A list containing service and additional hostnames
+        """
+        # The relation will always exist when this method is called
+        external_hostname = cast(str, get_config(self, "external-hostname"))
+        return external_hostname
+
+    def _are_relations_ready(self) -> bool:
+        """Check if required relations are ready.
+
+        Returns:
+            Whether required relations are ready and execution should continue.
+        """
+        return self._tls.get_tls_relation() is not None
+
+    def _reconcile(self) -> None:
+        """Reconcile charm status based on configuration and integrations."""
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            self.unit.status = BlockedStatus("Waiting for TLS.")
+            return
+        self.unit.status = ActiveStatus()
+
+    def _on_config_changed(self, _: Any) -> None:
+        """Handle the config-changed event."""
+        self._reconcile()
+
+    def _on_start(self, _: Any) -> None:
+        """Handle the start event."""
+        self._reconcile()
+
+    def _on_get_certificate_action(self, event: ActionEvent) -> None:
+        """Triggered when users run the `get-certificate` Juju action.
+
+        Args:
+            event: Juju event
+        """
+        hostname = event.params["hostname"]
+        tls_certificates_relation = self._tls.get_tls_relation()
+        if not tls_certificates_relation:
+            event.fail("Certificates relation not created.")
+            return
 
-    def _on_httpbin_pebble_ready(self, event: ops.PebbleReadyEvent) -> None:
-        """Define and start a workload using the Pebble API.
+        tls_rel_data = tls_certificates_relation.data[self.app]
+        if not tls_rel_data.get(f"certificate-{hostname}"):
+            event.fail("Certificate not available")
+            return
 
-        Change this example to suit your needs. You'll need to specify the right entrypoint and
-        environment configuration for your specific workload.
+        event.set_results(
+            {
+                f"certificate-{hostname}": tls_rel_data[f"certificate-{hostname}"],
```

Comment:
Minor: check if the relation data exists before setting the results?

Reply:
That's a good point, thanks! I updated the check to check for all relevant keys in the relation data. I also updated to using `.get()` to avoid KeyError. Will update this on the nginx ingress integrator side as well